clippy_lints/src/utils/conf.rs

```diff
@@ -102,11 +102,38 @@ macro_rules! define_Conf {
         $(#[conf_deprecated($dep:literal, $new_conf:ident)])?
         ($name:ident: $ty:ty = $default:expr),
     )*) => {
-        /// Clippy lint configuration
+        use std::sync::OnceLock;
+        pub static CONF_INSTANCE: OnceLock<Conf> = OnceLock::new();
+
+        /// Clippy lint configuration database.
+        #[derive(Debug, Clone)]
         pub struct Conf {
+            // TODO: make it private once migration completed
+            // to restrict access only from getter method
             $($(#[doc = $doc])+ pub $name: $ty,)*
         }
 
+        // Here we implement public getter with access checker.
+        impl Conf {
+            pub fn instance() -> &'static Self {
+                CONF_INSTANCE.get().expect("CONF_INSTANCE is not initialized")
+            }
+
+            $($(#[doc = $doc])+ pub fn $name(&self) -> $ty {
+                // Assume that this $name method is called from `check_(stmt|expr|...)` method defined in lints.
+                // Here we might need to implement validation to restrict call these getters unless
+                // the caller lint registers config properly.
+                // To implement validation:
+                // - A: caller must pass its names to this `$name(lint names)` method and this method
+                //      checks if given lint names is declared in `define_Conf!` macro.
+                // - B: define macro to implement wrapper struct of `Conf` for each lint struct.
```

Comment:
plan B might be impossible AFAIK

```rust
#[proc_macro]
pub fn impl_lint_conf(input: TokenStream) -> TokenStream {
    let LintConfig { lint_name, config } = parse_macro_input!(input as LintConfig);

    let config_struct_name = format_ident!("{lint_name}Config");

    // let getters  = config.iter().map(|config_name| {
    //     // TODO: acquire return type of #config_name...
    //     // EDIT: Since proc_macro runs before its compilation so it cannot acquire
    //     // type information in proc_macro definition
    //     quote! {
    //         pub fn #config_name()
    //     }
    // }).collect();

    let output = quote! {
        impl #lint_name {
            struct #config_struct_name {
                // pub #(#config)
            }

            impl #config_struct_name {
                pub fn new() -> Self {
                    #config_struct_name
                }
            }

            pub fn conf() -> #config_struct_name {
                #config_struct_name::new()
            }
        }
    };

    TokenStream::from(output)
}
```